### ISSUE TITLE
could not install with git http

### ISSUE THREAD

**Issue by alexey-shaposhnikov** [npm/npm#8636]
Hi dear developers. Thank you for the awesome npm.

I have a private repo on gitlab were available only http(not httpS). I haven't admin permissions. I want to install package with the following command
``` 
npm install git+http://gitlab.myawesome.com/project/name.git
```

npm debug log
```
cat npm-debug.log 
0 info it worked if it ends with ok
1 verbose cli [ 'node',
1 verbose cli   '/usr/local/bin/npm',
1 verbose cli   'install',
1 verbose cli   'git+http://gitlab.myawesome.com/project/name.git' ]
2 info using npm@2.11.3
3 info using node@v0.12.0
4 verbose node symlink /usr/local/bin/node
5 verbose install initial load of /Users/alexey/test/package.json
6 verbose readDependencies loading dependencies from /Users/alexey/test/package.json
7 silly cache add args [ 'git+http://gitlab.myawesome.com/project/name.git',
7 silly cache add   null ]
8 verbose cache add spec git+http://gitlab.myawesome.com/project/name.git
9 silly cache add parsed spec { raw: 'git+http://gitlab.myawesome.com/project/name.git,
9 silly cache add   scope: null,
9 silly cache add   name: null,
9 silly cache add   rawSpec: 'git+http://gitlab.myawesome.com/project/name.git,
9 silly cache add   spec: 'http://gitlab.myawesome.com/project/name.git,
9 silly cache add   type: 'git' }
10 verbose addRemoteGit caching git+http://gitlab.myawesome.com/project/name.git
11 silly tryClone cloning git+http://gitlab.myawesome.com/project/name.git via git+http://gitlab.myawesome.com/project/name.git
12 verbose tryClone some-name-git-1d546686 not in flight; caching
13 info git [ 'clone',
13 info git   '--template=/Users/alexey/.npm/_git-remotes/_templates',
13 info git   '--mirror',
13 info git   'http://gitlab.myawesome.com/project/name.git',
13 info git   '/Users/alexey/.npm/_git-remotes/name-git-1d546686' ]
14 error git clone --template=/Users/alexey/.npm/_git-remotes/_templates --mirror http://gitlab.myawesome.com/project/name.git /Users/alexey/.npm/_git-remotes/name-git-1d546686: Cloning into bare repository '/Users/alexey/.npm/_git-remotes/name-git-1d546686'...
14 error git clone --template=/Users/alexey/.npm/_git-remotes/_templates --mirror http://gitlab.myawesome.com/project/name.git /Users/alexey/.npm/_git-remotes/name-git-1d546686: fatal: Authentication failed for 'http://gitlab.myawesome.com/project/name.git'
15 verbose stack Error: Command failed: git clone --template=/Users/alexey/.npm/_git-remotes/_templates --mirror http://gitlab.myawesome.com/project/name.git /Users/alexey/.npm/_git-remotes/name-git-1d546686
15 verbose stack Cloning into bare repository '/Users/alexey/.npm/_git-remotes/name-git-1d546686'...
15 verbose stack fatal: Authentication failed for 'http://gitlab.myawesome.com/project/name.git'
15 verbose stack
15 verbose stack     at ChildProcess.exithandler (child_process.js:744:12)
15 verbose stack     at ChildProcess.emit (events.js:110:17)
15 verbose stack     at maybeClose (child_process.js:1008:16)
15 verbose stack     at Socket.<anonymous> (child_process.js:1176:11)
15 verbose stack     at Socket.emit (events.js:107:17)
15 verbose stack     at Pipe.close (net.js:476:12)
16 verbose cwd /Users/alexey/test
17 error Darwin 14.0.0
18 error argv "node" "/usr/local/bin/npm" "install" "git+http://gitlab.myawesome.com/project/name.git"
19 error node v0.12.0
20 error npm  v2.11.3
21 error code 128
22 error Command failed: git clone --template=/Users/alexey/.npm/_git-remotes/_templates --mirror http://gitlab.myawesome.com/project/name.git /Users/alexey/.npm/_git-remotes/name-git-1d546686
22 error Cloning into bare repository '/Users/alexey/.npm/_git-remotes/name-git-1d546686'...
22 error fatal: Authentication failed for 'http://gitlab.myawesome.com/project/name.git'
23 error If you need help, you may report this error at:
23 error     <https://github.com/npm/npm/issues>
24 verbose exit [ 1, true ]
```

The next command works properly but I really don't want to add my username and password into package.json

``` 
npm install git+http://username:password@gitlab.myawesome.com/project/name.git
```

The question: **how can I install pkg via http?**

**Comment by WORMSS** [comment]
Easiest way is to downgrade to npm 2.7.3, after that point, they broken username/password input and refuse to fix.

**Comment by othiym23** [comment]
@WORMSS can you clarify what you believe to be broken? The OP is asking for something new, which is a way to do authentication to git servers over HTTP[S] _without_ embedding the credentials in the URL. As far as I know, npm has never supported that (although it would be more secure if it did).

Also, I believe embedding the credentials into the URL does work (again), and if not, that's something we should fix.

**Comment by WORMSS** [comment]
You used to be able to do `npm install git+https://github.com/User/Repo.git`
Which would then ask for a username and password.
Then during all of your "Fixing" it now no longer asks for username/password to be entered, and just fails as no credentials.

This is your own words in **2.8.0** release notes:
`No longer prompt for credentials -- it didn't work right with the spinner, and wasn't guaranteed to work anyway. We may experiment with doing this a better way in the future. Users can override this by setting GIT_ASKPASS in their environment if they want to experiment with interactive cloning, but should also set --no-spin on the npm command line (or run npm config set spin=false).`

And you guys have not implemented a `npm install --no-spin --git-askpass https://github.com/User/Repo.git`
So I cannot just give out a simple command to people to install a private organisation module.
The `--no-spin` works, (as always).

I even tried releasing a .bat file (windows) to my co-workers, that would set the environment variable temporarily and then call `npm install ....`, but somewhere in the call chain, either the bat file calling npm or npm calling git, it lost access to that environment variable.

I didn't see him asking for a new feature, I just seen him ask for **any** way to pass the login credentials  that does not involve putting them in the url.

This was one of the problems I had as each user has their own username/password to the github organisation so I could not put them in up front. It used to ask and work as expected 2.7.3, so this is what I advise people to do.

- Colin.

**Comment by othiym23** [comment]
Thanks for the clarification, @WORMSS! I see the connection between this and the original post now. So, part of the reason we changed that behavior was the UI kludge it exposed, as described above. The _primary_ reason is that it made it impossible for users cloning via SSH to get their SSH keys used under certain circumstances. However, part of it was that that way of handling basic HTTP[S] authentication made it impossible to do unattended installs from git repositories that were specified that way, and that's what I meant when I discussed experimenting with a better way.

We have a better solution now for SSH (using a custom `GIT_ASKPASS`), and it would be nice to have something similar for HTTP. Figuring out how to do that is the next step to improving how npm handles these. I don't see us returning to just silently passing through the prompt, because it was a horrible user experience.

**Comment by benoror** [comment]
+1 for getting back some sort-of interactive password input

**Comment by leonexis** [comment]
+1 This issue still exists in 2.14.0. I still don't understand why _useful_ features were removed just to have a fancy spinner. This breaks simple manual deployment of internal modules and I'm not switching to your commercial service (possible motivation for the change?). Some of us require better control over who has access to our source code and using a third party service is not an option, which is why we use privately-served git repositories in the first place.

**Comment by benoror** [comment]
+2 Totally agree with @leonexis

**Comment by alexey-shaposhnikov** [comment]
well, hope that the feature will be added soon
thanks
![wait-meme-cool-hd](https://cloud.githubusercontent.com/assets/2097562/9519753/e9dfe0b0-4cda-11e5-971e-fd0849112009.jpg)

**Comment by JosephineJ** [comment]
I fully agree with @leonoxis and @WORMSS. Would really love to either have this feature of installing private git repos via a simple npm install (re)added or to at least include a more descriptive error message if install fails in this use case. I'm rather new to npm and it actually took me quite some time to figure that problem out and find this issue here. Looking forward to your changes in the upcoming releases of npm!

Many thanks to @alexey-shaposhnikov - opening this issue helped me out a lot!

**Comment by fpauser** [comment]
Same problem here... https-auth urgently needed

**Comment by mikaelec** [comment]
:+1:

**Comment by mikaelec** [comment]
I would also like git https authentication re-added somehow.

**Comment by lokhmakov** [comment]
Same, bitbucket https auth needed

**Comment by jonaskello** [comment]
+1 for https auth

**Comment by lucasfeliciano** [comment]
+1 for https auth

**Comment by othiym23** [comment]
The CLI team agrees that forcing the credentials for a protected resource to be included in the URL is both awkward and insecure. This is especially true for hosted Git repos, because they’re so commonly used during package development. The most sensible solution would be for npm to integrate with Git’s credentials management tools directly, but this is kind of tricky because of npm’s need to work with a wide variety of Git setups on multiple platforms in a seamless way.

In the long term, the user community would benefit if npm has generic credential management, but in the interest of getting to this soon, we think we’re probably going to start with Git. There is some design work that needs to be done before implementation (so this isn’t ready for community patches yet), but this is something the team intends to tackle within the next 6-12 months as part of our ongoing work in improving how npm works with Git.

**Comment by WORMSS** [comment]
Thank you for keeping us uptodate with this issue.

**Comment by igabesz** [comment]
Maybe this is a superset of #11567 and the related issues? If yes, I'm really looking forward to it.

**Comment by blowsie** [comment]
Thanks for all of this useful information.

Just for clarity for myself and others; is the only solution to this to roll back NPM to v2.7.3?

**Comment by WORMSS** [comment]
Not the **only** solution.

1. Due to latest version's of nodejs for windows comes with a later version of npm and sounds silly telling people to downgrade. We have been sending out a zip file to our developers which as a install.bat and install.sh which does nothing more than `npm install -g .` (dot being current directory). The con to this is any change to the module requires sending out another zip file and people installing again, but it happens very rarely it's not become too much of an issue.

2. Pay for private modules at npmjs.

3. Run a private registry. Quick Google shows http://thejackalofjavascript.com/maintaining-a-private-npm-registry/ I've not read it, but should give the the idea.

**Comment by gudmundur** [comment]
Since Git uses curl under the hood. It's possible to set up a `~/.netrc` file with the credentials.

**Comment by Hypfer** [comment]
So I just ran into this two year old issue and there still is no solution? Wow.

**Comment by alexey-sh** [comment]
@Hypfer use http**s**, Luke

**Comment by WORMSS** [comment]
>  the team intends to tackle within the next 6-12 months as part of our ongoing work in improving how npm works with Git.

@othiym23 any update of this?

**Comment by rdesoky** [comment]
This worked for me (first two steps are required once)
git config --global credential.helper cache
git ls-remote https://myorggithub.com/user/repo
npm install git+https://myorggithub.com/user/repo

**Comment by claytongulick** [comment]
On windows, upgrading to the latest git (2.15.0) and installing the Git Credential Cache for windows solved the issue for me. I cached by credentials as suggested by @rdesoky above.

**Comment by edhemphill** [comment]
Why did I spend 30 minutes having to Google this???

setting `git --config credentials.helper cache`  and seeding it with a user / pass worked for me on Ubuntu, with a `https://` style npm path.

That being said I'd really recommend the `npm` team fix this to prompt if no password. The spinner stuff does not make the product any better, but private repos are super common and tool appears quite broken without it prompting.
